### ISSUE TITLE
Shoutcast stream causing mistaken ID3v2::FrameFactory call to be very slow

### ISSUE THREAD

**Issue by DaveTBlake** [taglib/taglib#792]
Some internet radio streams end ".mp3", and thus have mistakenly passed them to TagLib as if they were tagged with ID3 metadata. The stream does contain mp3 data, but with its own Shoutcast protocol metadata. 
With Taglib v1.11 the call
` file = mpegFile = new MPEG::File(stream, ID3v2::FrameFactory::instance());`
can take 10mins or more.

With v1.09 for the same station this delay was only 3s. Delay varies with stream.

It would be helpful if TagLib could identify there is no ID3 header and fail gracefully, instead of being so slow and then returning succussfully but without any tag data.

Any suggestions about this? Yes I know not calling Taglib with a radio stream.
But how does TagLib handle the factory being called for the wrong header format?

**Comment by TsudaKageyu** [comment]
Hi @DaveTBlake,
Could you provide us with some sample file that takes long time to read? If something goes wrong with TagLib, we will fix it.

**Comment by DaveTBlake** [comment]
Thanks for looking into this issue.  This station is particularly slow 
​http://94.199.183.186/jazzy.mp3 
but there are others.  And v1.11 is much slower than v1.09 which still had a delay but only seconds.

In comparison http://icyrelay.181.fm/181-vibe_128k.mp3 only briefly hesitates, so TagLib can cope with some Shoutcast streams better than others.

**Comment by TsudaKageyu** [comment]
Thanks! It was a great help.
I have located the cause of the problem and will fix it in a few days.

**Comment by DaveTBlake** [comment]
Great news, thanks

**Comment by TsudaKageyu** [comment]
I have made a commit to fix this issue. TagLib now reads jazzy.mp3 almost without delay.

**Comment by DaveTBlake** [comment]
Thanks for this, I have tested with various radio stations that were a problem and confirm this resolves the issue.

Would it be possible to do a maintenance release 1.11.2 that includes this fix? 

Since the behaviour seems to be a regression since v1.9 there could be others that benefit. Unfortunately some of the commits to master since 1.11.1 seem too risky for an interim build to be used in a Kodi formal release, and I would not know how to selectively pick commits with confidence. Hope that you can help.

**Comment by TsudaKageyu** [comment]
Right, I admit that it's a regression and it's good for us to do a maintenance release.

However, I have another question about this issue now.
The root cause of this issue is that TagLib looks for an ID3v2 tag until reaching the first valid MPEG frame (not only the beginning of a file). Most software assumes that an ID3v2 tag is always placed at the beginning of a file and they work well. 
Does TagLib really need to look not only the beginning of a file for an ID3v2 tag?

I'd like to ask other members' opinions.

**Comment by sbooth** [comment]
I agree it's probably rare for an ID3v2 tag not to be at the beginning of the file.  I read the informal ID3V2.4 spec (http://id3.org/id3v2.4.0-structure) and it says:

> 5.   Tag location
> 
>    The default location of an ID3v2 tag is prepended to the audio so
>    that players can benefit from the information when the data is
>    streamed. It is however possible to append the tag, or make a
>    prepend/append combination. When deciding upon where an unembedded
>    tag should be located, the following order of preference SHOULD be
>    considered.
>    
>      1. Prepend the tag.
> 
>      2. Prepend a tag with all vital information and add a second tag at 
>         the end of the file, before tags from other tagging systems. The
>         first tag is required to have a SEEK frame.
>       
>      3. Add a tag at the end of the file, before tags from other tagging
>         systems.
>       
>    In case 2 and 3 the tag can simply be appended if no other known tags
>    are present. The suggested method to find ID3v2 tags are:
>    
>      1. Look for a prepended tag using the pattern found in section 3.1.
> 
>      2. If a SEEK frame was found, use its values to guide further
>         searching.
> 
>      3. Look for a tag footer, scanning from the back of the file.
> 
>    For every new tag that is found, the old tag should be discarded
>    unless the update flag in the extended header (section 3.2) is set.
> 

so it seems that tags at the end of files are explicitly allowed.  Should TagLib handle this case?  I tend to think so.

**Comment by TsudaKageyu** [comment]
Except for MPEG, AIFF and WAV, TagLib assumes that an ID3v2 tag is always at the beginning of a file.
AIFF and WAV have their own "chunk" structure, so a tag is stored in a designated chunk.

**Comment by DaveTBlake** [comment]
@ TsudaKageyu what are your final thoughts on this?

Will TagLib continue to look for an ID3v2 tag until reaching the first valid MPEG frame (not only the beginning of a file)? Does this mean that Taglib should only be used to scan files of a **limited size**, so will have problems if passed something like an internet radio stream?

**Comment by TsudaKageyu** [comment]
Sorry, I mistakenly closed this issue.

Currently, (At least I believe) `MPEG::File` works fine as long as a valid MPEG stream is given. So I think we don't have a dire need to change the current behavior.

**Comment by sbooth** [comment]
@DaveTBlake Wouldn't an internet radio stream contain valid MPEG frames, just perhaps not initially aligned with the start of the buffer?  I think TagLib should handle this (but I haven't tried yet).

**Comment by TsudaKageyu** [comment]
A radio stream must have valid MPEG frames as long as the stream is MP3.
If we give a non-MPEG stream which has no ID3v2 tag to `MPEG::File`, it will take a long time since TagLib scan the entire stream.

**Comment by DaveTBlake** [comment]
Shoutcast content is mp3 yes, with its own header protocol, but it is continuous, not a finite file. 

TagLib 1.11.1 was not handling some Shoutcasts, but https://github.com/taglib/taglib/commit/d2e0e5522308ce4c905c205c7bf5972ea71cd6e5 fixed that.

But it has left me wondering if those of us using TagLib need to take more care over what we call TagLib with.

**Comment by DaveTBlake** [comment]
Just a little more on this issue.  Having similar hanging problens with ".acc" streams e.g.
http://icy-e-bz-06-cr.sharp-stream.com/planetrock.aac

Patch does not fix that

**Comment by ufleisch** [comment]
The example files no longer exist, but as the mentioned commit is present in the code, I assume that this is fixed. Please reopen if I am wrong.
